t1rocketemu/src/AXI4SlaveAgent.scala

```diff
@@ -58,189 +73,156 @@ class ReadPayload(length: Int, dataWidth: Int) extends Bundle {
 class AXI4SlaveAgent(parameter: AXI4SlaveAgentParameter)
     extends FixedIORawModule[AXI4SlaveAgentInterface](new AXI4SlaveAgentInterface(parameter)) {
   dontTouch(io)
+  val invClock = (~io.clock.asBool).asClock;
+
   io.channel match {
-    case channel: AXI4RWIrrevocableVerilog =>
-      new WriteManager(channel)
-      new ReadManager(channel)
-    case channel: AXI4ROIrrevocableVerilog =>
-      new ReadManager(channel)
-    case channel: AXI4WOIrrevocableVerilog =>
-      new WriteManager(channel)
+    case channel: AXI4RWIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4RWIrrevocable](
+        implicitly,
+        AXI4RWIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new ReadManager(view)
+      new WriteManager(view)
+    }
+    case channel: AXI4ROIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4ROIrrevocable](
+        implicitly,
+        AXI4ROIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new ReadManager(view)
+    }
+    case channel: AXI4WOIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4WOIrrevocable](
+        implicitly,
+        AXI4WOIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new WriteManager(view)
+    }
+  }
+
+  RawClockedVoidFunctionCall(s"axi_tick")(
+    io.clock,
+    true.B,
+    io.reset.asTypeOf(UInt(8.W))
+  )
+
+  /// Widen a wire to make all DPI calls' types uniform
+  def widen(wire: UInt, target: Int): UInt = {
+    require(wire.isWidthKnown)
+    require(wire.getWidth <= target)
+    wire.asTypeOf(UInt(target.W))
   }
 
-  private class WriteManager(
-    channel: AWChannel with AWFlowControl with WChannel with WFlowControl with BChannel with BFlowControl) {
+  private class WriteManager(channel: HasAW with HasW with HasB) {
     withClockAndReset(io.clock, io.reset) {
+      val awqueue = Module(new Queue(new AW(parameter.axiParameter), 2))
+      val wqueue  = Module(new Queue(new W(parameter.axiParameter), 2))
+      val bqueue  = Module(new Queue(new B(parameter.axiParameter), 2))
+
+      awqueue.io.enq <> channel.aw
+      wqueue.io.enq <> channel.w
+      bqueue.io.deq <> channel.b
+
+      // Invoke DPI at negedge
+      // NOTICE: this block CANNOT directly write any outside reg. Only write wires (e.g. here, only writes queue IO)
+      withClock(invClock) {
```

Comment:
`RawCLockedNonVoidFunctionCall` pass clock explicitly. No need to `withClock` here.

And please separate changes on AXI4SlaveAgent (from changes on rust code) to their own commits , with `[axi4] ...` commit message , to make the history cleaner.